statocles-site/blog/2024/12/16/pitch-detection/index.markdown

```diff
@@ -0,0 +1,243 @@
+---
+title: Day 16: Pitch detection
+disable_content_template: 1
+tags:
+    - signal processing
+    - phonetics
+    - speech
+author: José Joaquín Atria
+images:
+  banner:
+    src: 'banner.jpg'
+    alt: A close up of a person's ear
+    data:
+      attribution: |-
+        <a href="https://www.pexels.com/photo/close-up-photo-of-a-person-s-ear-8092973/">Close-Up Photo of a Person's Ear</a> Image Credit: Kaboompics.com, <a href="https://www.pexels.com/license/">free to use</a>
+data:
+  bio: jjatria
+  description: |
+    An explanation of (part of) Praat's pitch detection algorithm using PDL
+---
+
+In a previous life I spent most of my time (free and otherwise) thinking about
+phonetics and the analysis of speech sounds. Like most scientific research, it
+was slow and tedious and oddly rewarding.
+
+Back then I was already in a steady drift towards spending more time playing
+with computers than with my research, so a lot of my efforts went into
+studying [Praat](https://www.fon.hum.uva.nl/praat), a piece of software you've
+almost certainly never heard of, but that has probably been used in most of
+the phonetics research you've also never heard of.
+
+One of Praat's biggest claims to fame is its [pitch detection algorithm],
+which does a remarkably good job at estimating the [fundamental frequency]
+(or f₀) of speech sounds. This is essential for any kind of research on the
+intonation of spoken languages (called "prosody"), and it is difficult for
+primarily two reasons:
+
+[pitch detection algorithm]: https://www.fon.hum.uva.nl/praat/manual/FAQ__Pitch_analysis.html
+[fundamental frequency]: https://en.wikipedia.org/wiki/Fundamental_frequency
+
+1.  Speech sounds are made of a combination of several harmonic frequencies,
+    which makes the task not just about frequency detection, but about
+    detecting the right one.
+
+2.  Speech sounds change rapidly, so we have only a small amount of unstable
+    data for any frequency at any point in time.
+
+Praat's algorithm corrects for these by making a number of assumptions about
+the sounds it deals with, most of which are tied to the fact that speech
+sounds are generated by the human vocal tract. So, for example, while speech
+sounds change all the time, there is a limit to how quickly the vocal tract
+can make those changes, which means that on a small enough time scale, we
+should be able to treat speech sounds as if they were stable. Praat refers
+to these windows of analysis as "frames".
+
+This article (which is based on [a similar one] I wrote in 2012 using R) will
+go over the core part of the algorithm as described in [Boersma (1993)], and
+since this is the PDL Advent Calendar, we'll be using PDL for illustrating
+each step and generating the figures. I'll be showing the relevant bits of
+code at each step, but you can also download [a file with all of the code] in
+context.
+
+[boersma (1993)]: http://www.fon.hum.uva.nl/paul/papers/Proceedings_1993.pdf
```

Comment:
Can you confirm you've checked Statocles likes this formatting? See my notes on #38 for how to iterate.

Reply:
You mean the links with the implicit references, etc? Yeah, I confirmed it was working.

Incidentally, this is the command I was using to try it, which means it didn't mess with the directory I was currently in:

```
pushd $( git rev-parse --show-toplevel ); (cd statocles-site/; rm -rf .statocles; statocles build --date 2024-12-30); rm -rf advent/; cp -r statocles-site/.statocles/build/ advent/; popd
```